### ISSUE TITLE
C Tracking Issue

### ISSUE THREAD

**Issue by mgattozzi** [mgattozzi/rust-from-lang#14]
Article requests as well as planned articles will be tracked in a list here. Even if the article isn't listed here but you want to write about something that's showing off how to do C like things in Rust that's fine! Otherwise feel free to write an article from here. If you want a C based article please make a request here.

**Comment by jonhoo** [comment]
Suggestion: Pointers and references
Comparing C pointers to Rust references, including Rust's restrictions on ownership and mutability, `NULL`, pointer arithmetic and aliasing, etc.

**Comment by mgattozzi** [comment]
Absolutely. Also things like iterators vs for loops for accessing data in arrays for instance.

**Comment by Lokathor** [comment]
I've been writing a version of the [Handmade Hero](http://handmadehero.org) project in rust this month. So far I've seen that, in general, unoptimized rust performs a lot worse than unoptimized C. Yes, the can end up near the same speed with optimizations on, but new users could easily not know about the difference there, so I'd be sure to explain it, because speed is usually the only king with the C crowd.

In particular, `for` and `while` are essentially interchangeable in C, but in rust a `for` loop will run _much_ worse than a similar `while` loop unless you have your optimizations on, because of the iterator overhead doesn't go away until those higher optimization levels.

**Comment by mgattozzi** [comment]
I think this has a good overview about those iterator guts not being efficient:
https://medium.com/@veedrac/rust-is-slow-and-i-am-the-cure-32facc0fdcb#.69120rlm4

This might be a good thing to talk about if people do benchmarking stuff. If you turn the `release` flag on are the loops fairly equivalent in speed or is it some other flag?

**Comment by Lokathor** [comment]
Using `--release` just sets you to build using the `release` profile instead of the `dev` profile. The real effect is actually configured by your cargo manifest file, specifically the four [profiles sections](http://doc.crates.io/manifest.html#the-profile-sections), where you can specify a number of build options. Note that all the entries listed on that page are the _default_ entries for that profile type, so if you don't want a change you don't need to list that part of the entry at all. Which is why not having any profile entries also works well.

What speeds up iterators and gets them to compile down is having a higher optimization level. I've been told that "1 helps, more is better", but I don't know the exact details there. By default, `release` and `bench` are opt-level 3 (but LTO off), which is why their loops are so much snappier. You can raise the opt-level on dev and testing if you like, but with a corresponding effect on compile times of course. Some projects are so loop heavy that you want something like opt-level 1 even just for debug or the UI can't stay responsive even in demo mode.
